### PR TITLE
Add distance transform implementation

### PR DESCRIPTION
**Pull request by aaroswings**
## Changes

Added "Fast Cascaded Convolutional Distance Transform",  an "ad-hoc" differentiable loose approximation of a 2D distance transform for images. The feature is motivated by the potential for its use as a regression loss in image segmentation model development. Tests are also added in the test_contrib file.

### References:

Karam, C.; Sugimoto, K.; Hirakawa, K., "Fast Convolutional
Distance Transform," Signal Processing Letters (SPL), 2019 IEEE Journal.

Pham et al, "A Differentiable Convolutional Distance
Transform Layer for Improved Image Segmentation"     Pattern Recognition, 2021 Conference Proceedings

#### Type of change
- [x] 🔬 New feature (non-breaking change which adds functionality)

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

File: test/test_contrib.py [hunk 1]
```diff
@@ -371,8 +371,53 @@ def test_exception(self, device, dtype):
...
+        input1 = torch.rand(B, C, H, W, device=device, dtype=dtype)
+        distance_transformer = kornia.contrib.ConvDistanceTransform(kernel_size)
+
+        output1 = distance_transformer(input1)
+        output2 = kornia.contrib.conv_distance_transform(input1, kernel_size)
+
+        assert isinstance(output1, torch.Tensor)
+        assert isinstance(output2, torch.Tensor)
+        assert output1.shape == input1.shape
+
```
Comment (edgarriba) [inline test/test_contrib.py:387]: I miss a couple of `test_value` to verify results vs expected. We also use this tests to showcase as a small snippet in the docstrings.

File: test/test_contrib.py [hunk 2]
```diff
@@ -371,8 +371,53 @@ def test_exception(self, device, dtype):
...
+    def test_module(self, device, dtype):
+        B, C, H, W = 1, 3, 99, 100
+        input = torch.rand(B, C, H, W, device=device, dtype=dtype)
+        distance_transformer = kornia.contrib.ConvDistanceTransform()
```
Comment (edgarriba) [inline test/test_contrib.py:391]:
```suggestion
        distance_transformer = kornia.contrib.ConvDistanceTransform().to(device, dtype)
```

File: test/test_contrib.py [hunk 3]
```diff
@@ -371,8 +371,53 @@ def test_exception(self, device, dtype):
...
+            kornia.contrib.distance_transform.conv_distance_transform(input1, 4)
+
+        # Invalid input dimensions
+        with pytest.raises(ValueError):
+            kornia.contrib.distance_transform.conv_distance_transform(input2)
+
+        # Invalid input type
+        with pytest.raises(TypeError):
+            kornia.contrib.distance_transform.conv_distance_transform(None)
+
```
Comment (edgarriba) [inline test/test_contrib.py:418]: could you also ass a `test_gradcheck` ?

**Review by edgarriba — COMMENTED** [review]
Not bad ! just some considerations -- @ducha-aiki please review as well

File: Code test.ipynb [hunk 4]
```diff
@@ -0,0 +1,407 @@
+{
```
[the remark this answers is not in the document]
Comment (aaroswings) [reply Code test.ipynb]: I'm removing the notebooks (including them was actually an accident, oops!)

File: kornia/contrib/distance_transform.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+    gridx, gridy = torch.meshgrid(grid_range, grid_range)
+    gridx = gridx - math.floor(kernel_size / 2)
+    gridy = gridy - math.floor(kernel_size / 2)
+
+    kernel = torch.hypot(gridx, gridy)
+    kernel = torch.exp(kernel / h)
+
+    # for BCHW tensors
+    kernel = torch.unsqueeze(kernel, 0)
```
[the remark this answers is not in the document]
Comment (aaroswings) [reply kornia/contrib/distance_transform.py:22]: Would you mind expanding on this point?

[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/contrib/distance_transform.py:22]: just to fuse 3 to 1 lines

File: kornia/contrib/distance_transform.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+import kornia
+
+
+def make_cdt_kernel(
+    kernel_size: int,
+) -> torch.Tensor:
+    # Value of h is derived from the parameters and reference code given by the authors who proposed the algorithm.
+    h = -0.35
+
+    grid = kornia.utils.create_meshgrid(kernel_size, kernel_size, normalized_coordinates=False)
```
Comment (edgarriba) [inline kornia/contrib/distance_transform.py:16]: for internal development and avoid circular package dependencies we tend to do
```suggestion
    grid = create_meshgrid(kernel_size, kernel_size, normalized_coordinates=False)

```
then `from kornia.utils import create_meshgrid`

File: kornia/contrib/distance_transform.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+import torch.nn.functional as F
+
+import kornia
+
+
+def make_cdt_kernel(
+    kernel_size: int,
+) -> torch.Tensor:
+    # Value of h is derived from the parameters and reference code given by the authors who proposed the algorithm.
+    h = -0.35
```
Comment (edgarriba) [inline kornia/contrib/distance_transform.py:14]:
```suggestion
    h: float = -0.35

```

File: kornia/contrib/distance_transform.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    The transformation is applied independently across the channel dimension of the images.
+
+
+    Args:
+        image: Image with shape :math:`(B,C,H,W)`.
+        kernel_size: size of the convolution kernel. Larger kernels are more accurate but less numerically stable.
+
+    Returns:
+        tensor with shape :math:`(B,C,H,W)`.
+
```
Comment (edgarriba) [inline kornia/contrib/distance_transform.py:45]: can you add here a small example to showcase the usage?

File: kornia/contrib/distance_transform.py (hunk 7 above)
Comment (edgarriba) [reply kornia/contrib/distance_transform.py:14]: we use also this value in the main function, possibly we could pass as function parameter

File: kornia/contrib/distance_transform.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    for i in range(n_iters):
+        cdt = F.conv2d(boundary, kernel, padding='same')
+        cdt = -0.35 * torch.log(cdt)
+
+        # We are calculating log(0) above.
+        cdt = torch.nan_to_num(cdt, posinf=0.0)
+
+        mask = cdt > 0
+        if mask.sum() == 0:
```
Comment (edgarriba) [inline kornia/contrib/distance_transform.py:79]: not for now but can we find an alternative to avoid this extra complexity every iteration ?

File: kornia/contrib/distance_transform.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        boundary[mask] = 1
+
+    # View channels in the channel dimension, if they were added to batch dimension during transform.
+    if image.shape[1] > 1:
+        out = out.view(image.shape)
+
+    return out.view_as(image)
+
+
+class ConvDistanceTransform(nn.Module):
```
Comment (edgarriba) [inline kornia/contrib/distance_transform.py:93]:
```suggestion
class DistanceTransform(nn.Module):

```

File: kornia/contrib/distance_transform.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    Args:
+        kernel_size: size of the convolution kernel. Larger kernels are more accurate but less numerically stable.
+
+    """
+    def __init__(
+        self,
+        kernel_size: int = 7
+    ):
+        super().__init__()
+        if kernel_size % 2 == 0:
```
Comment (edgarriba) [inline kornia/contrib/distance_transform.py:105]: this will be raised inside the functional right ?

File: test/test_contrib.py [hunk 12]
```diff
@@ -371,8 +371,71 @@ def test_exception(self, device, dtype):
...
+        # Invalid input dimensions
+        with pytest.raises(ValueError):
+            kornia.contrib.distance_transform(input2)
+
+        # Invalid input type
+        with pytest.raises(TypeError):
+            kornia.contrib.distance_transform(None)
+
+    def test_value(self, device, dtype):
```
Comment (edgarriba) [inline test/test_contrib.py:419]: cool -- use this as the example in docstrings

**Review by edgarriba — COMMENTED** [review]
just minor changes and good to go. One more thing, please add to documentation: https://github.com/kornia/kornia/blob/master/docs/source/contrib.rst#image-segmentation

File: kornia/contrib/distance_transform.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+
+        # We are calculating log(0) above.
+        cdt = torch.nan_to_num(cdt, posinf=0.0)
+
+        mask = cdt > 0
+        if mask.sum() == 0:
+            break
+
+        offset: int = i * kernel_size / 2
+        out[mask] += offset + cdt[mask]
```
Comment (edgarriba) [inline kornia/contrib/distance_transform.py:83]:
```suggestion
        out[mask] = out[mask] + offset + cdt[mask]

```
@aaroswings I've been trying this code and this gives problems during back-propagation since it's an inplace ooperation

File: kornia/contrib/distance_transform.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+        boundary[mask] = 1
```
Comment (edgarriba) [inline kornia/contrib/distance_transform.py:84]: this sometimes throws and error for being inplace

File: kornia/contrib/distance_transform.py (hunk 11 above)
Comment (aaroswings) [reply kornia/contrib/distance_transform.py:105]: Understandable, I'll take it out of those lines.

File: kornia/contrib/distance_transform.py (hunk 13 above)
Comment (aaroswings) [reply kornia/contrib/distance_transform.py:83]: I'm changing it to take out the inplace operation, let's see if this solves the issue.

File: kornia/contrib/distance_transform.py (hunk 7 above)
Comment (aaroswings) [reply kornia/contrib/distance_transform.py]: Adding it in as one

File: kornia/contrib/distance_transform.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+    Args:
+        image: Image with shape :math:`(B,C,H,W)`.
+        kernel_size: size of the convolution kernel.
+        h: value that influence the approximation of the min function.
+
+    Returns:
+        tensor with shape :math:`(B,C,H,W)`.
+
+    Example:
+        >>> tensor = tensor = torch.zeros(1, 1, 5, 5)
```
Comment (edgarriba) [inline kornia/contrib/distance_transform.py:30]:
```suggestion
        >>> tensor = torch.zeros(1, 1, 5, 5)

```

File: kornia/contrib/distance_transform.py [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+    if not len(image.shape) == 4:
+        raise ValueError(f"Invalid image shape, we expect BxCxHxW. Got: {image.shape}")
+
+    if kernel_size % 2 == 0:
+        raise ValueError("Kernel size must be an odd number.")
+
+    # n_iters is set such that the DT will be able to propagate from any corner of the image to its far,
+    # diagonally opposite corner
+    n_iters: int = math.ceil(max(image.shape[2], image.shape[3]) / math.floor(kernel_size / 2))
+    grid = create_meshgrid(kernel_size, kernel_size, normalized_coordinates=False)
```
Comment (edgarriba) [inline kornia/contrib/distance_transform.py:46]:
```suggestion
    grid = create_meshgrid(kernel_size, kernel_size, normalized_coordinates=False, device=image.device, dtype=image.dtype)

```

File: kornia/contrib/distance_transform.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+    grid -= math.floor(kernel_size / 2)
+    kernel = torch.hypot(grid[0, :, :, 0], grid[0, :, :, 1])
+    kernel = torch.exp(kernel / -h).unsqueeze(0).unsqueeze(0)
+    kernel = kernel.to(image)
```
Comment (edgarriba) [inline kornia/contrib/distance_transform.py:50]:
```suggestion

```

File: kornia/contrib/distance_transform.py [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+import torch.nn as nn
+import torch.nn.functional as F
+
+from kornia.utils import create_meshgrid
+
+
+def distance_transform(
+    image: torch.Tensor,
+    kernel_size: int = 3,
+    h: float = .35
```
Comment (edgarriba) [inline kornia/contrib/distance_transform.py:13]:
```suggestion
    h: float = 0.35

```

File: kornia/contrib/distance_transform.py [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+    kernel = kernel.to(image)
+
+    out = torch.zeros_like(image)
+
+    # It is possible to avoid cloning the image if boundary = image, but this would require modifying the image tensor.
+    boundary = image.clone()
+
+    for i in range(n_iters):
+        cdt = F.conv2d(boundary, kernel, padding='same')
+        cdt = -0.35 * torch.log(cdt)
```
Comment (edgarriba) [inline kornia/contrib/distance_transform.py:59]:
```suggestion
        cdt = -h* torch.log(cdt)

```

File: kornia/contrib/distance_transform.py [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+        h: float = .35
+    ):
+        super().__init__()
+        self.kernel_size = kernel_size
+        self.h = h
+
+    def forward(self, image: torch.Tensor) -> torch.Tensor:
+        # If images have multiple channels, view the channels in the batch dimension to match kernel shape.
+        if image.shape[1] > 1:
+            image_in = image.view(-1, 1, image.shape[-2], image.shape[-1])
```
Comment (edgarriba) [inline kornia/contrib/distance_transform.py:95]: Shouldn’t happen this in the functional as well?

Comment (aaroswings) [reply kornia/contrib/distance_transform.py:96]: I moved it out of the function into the module - would it be better to happen in both places or would that be redundant?

Comment (edgarriba) [reply kornia/contrib/distance_transform.py:96]: yes, i think better in just one place. @justanhduc was working on a small project to automate that, but we encountered some issue with scripting. So, I'd say that we perform that manually for now ?

File: kornia/contrib/distance_transform.py [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+    kernel = torch.hypot(grid[0, :, :, 0], grid[0, :, :, 1])
+    kernel = torch.exp(kernel / -h).unsqueeze(0).unsqueeze(0)
+
+    out = torch.zeros_like(image)
+
+    # It is possible to avoid cloning the image if boundary = image, but this would require modifying the image tensor.
+    boundary = image.clone()
+
+    for i in range(n_iters):
+        cdt = F.conv2d(boundary, kernel, padding='same')
```
Comment (edgarriba) [inline kornia/contrib/distance_transform.py:59]: `padding=='same'` seems that is only supported in pytorch > 1.9
https://discuss.pytorch.org/t/typeerror-conv2d-argument-padding-position-5-must-be-tuple-of-ints-not-str/130513/2

one suggestion is to use `kornia.filters.filter2d` where we handle this already /cc @ducha-aiki 
https://kornia.readthedocs.io/en/latest/filters.html?highlight=filter#kornia.filters.filter2d

I see that for this same reason the numerical test are not consistent

Comment (aaroswings) [reply kornia/contrib/distance_transform.py]: I changed it to use filter2d. (Apologies for the wait on responding to these changes, life's been a bit busy!)

File: kornia/contrib/distance_transform.py (hunk 20 above)
Comment (aaroswings) [reply kornia/contrib/distance_transform.py:96]: Ok. Is this alright as-is? Or should it be inside the function rather than the module?

File: test/test_contrib.py [hunk 22]
```diff
@@ -371,8 +371,73 @@ def test_exception(self, device, dtype):
...
+        input1 = torch.zeros(B, C, H, W, device=device, dtype=dtype)
+        input1[:, :, 1, 1] = 1.0
+        expected_output1 = torch.tensor([[[
+            [1.4142135382, 1.0000000000, 1.4142135382, 2.2360680103],
+            [1.0000000000, 0.0000000000, 1.0000000000, 2.0000000000],
+            [1.4142135382, 1.0000000000, 1.4142135382, 2.2360680103],
+            [2.2360680103, 2.0000000000, 2.2360680103, 2.8284270763]
+        ]]]).to(input1)
+        output1 = kornia.contrib.distance_transform(input1, kernel_size, h)
+        assert (expected_output1 == output1).all()
```
Comment (edgarriba) [inline test/test_contrib.py:432]:
```suggestion
        assert_close(expected_output1, output1)
```

File: test/test_contrib.py [hunk 23]
```diff
@@ -371,8 +371,73 @@ def test_exception(self, device, dtype):
...
+        kernel_size = 7
+        h = 0.35
+        input1 = torch.zeros(B, C, H, W, device=device, dtype=dtype)
+        input1[:, :, 1, 1] = 1.0
+        expected_output1 = torch.tensor([[[
+            [1.4142135382, 1.0000000000, 1.4142135382, 2.2360680103],
+            [1.0000000000, 0.0000000000, 1.0000000000, 2.0000000000],
+            [1.4142135382, 1.0000000000, 1.4142135382, 2.2360680103],
+            [2.2360680103, 2.0000000000, 2.2360680103, 2.8284270763]
+        ]]]).to(input1)
```
Comment (edgarriba) [inline test/test_contrib.py:430]:
```suggestion
        ]]], device=device, dtype=dtype)
```

**Review by edgarriba — APPROVED** [review]

**Comment by edgarriba** [comment]
@aaroswings I thing this is ready to go. We should plan a couple of example.
1) A basic colab just showing the result of applying the distance transform to a batch of images
2) Your cool example showing differentiability

I can take 1)
